### PR TITLE
refactor: add input validations when setting data in `LSP6SetDataModule`

### PR DESCRIPTION
**Pull request by b00ste**
# What does this PR introduce?

## ♻️ Refactor

Add input validations for the data values of specific data keys.

- [ ] Wrote Tests
- [ ] Wrote & Generated Documentation (readme/natspec/dodoc)
- [ ] Ran `npm run lint` && `npm run lint:solidity` (solhint)
- [ ] Ran `npm run format` (prettier)
- [ ] Ran `npm run build`
- [ ] Ran `npm run test`

### CODE REVIEW

**Comment by Hugoo** [comment]
Task linked: [DEV-7685 [C4#115] Missing input validations](https://app.clickup.com/t/2645698/DEV-7685)

File: contracts/LSP6KeyManager/LSP6Errors.sol [hunk 1]
```diff
@@ -242,6 +242,25 @@ error RelayCallBeforeStartTime();
...
+ * @dev Reverts when the address of the Key Manager is being set as extensions for lsp20 functions
  */
 error KeyManagerCannotBeSetAsExtensionForLSP20Functions();
+
+/**
+ * @notice Data value: `dataValue` length is different from `zeroValueRequiredLength` and `normalValueRequiredLength`.
+ *
+ * @dev Reverts when the data value length is not one of the required lengths for the specific data key.
+ *
+ * @param dataKey The data key that requires a specific length for the `dataValue`.
```
Comment (CJ42) [inline contracts/LSP6KeyManager/LSP6Errors.sol:256]: Suggested changes:

```suggestion
 * @param dataKey The data key associated with the invalid `dataValue` that was set.
```

File: contracts/LSP6KeyManager/LSP6Errors.sol [hunk 2]
```diff
@@ -242,6 +242,25 @@ error RelayCallBeforeStartTime();
...
+ * @param dataValue The data value that must have a specific length.
```
Comment (CJ42) [inline contracts/LSP6KeyManager/LSP6Errors.sol:257]:
```suggestion
 * @param dataValue The data value that has an invalid length.
```

File: contracts/LSP6KeyManager/LSP6Errors.sol [hunk 3]
```diff
@@ -242,6 +242,25 @@ error RelayCallBeforeStartTime();
...
+ * @param zeroValueRequiredLength The `dataValue` length required for any normal value set under the `dataKey`.
```
Comment (CJ42) [inline contracts/LSP6KeyManager/LSP6Errors.sol:258]: I am not sure I understand this parameter name. What does the `zeroValue` mean in the parameter name?

File: contracts/LSP6KeyManager/LSP6Modules/LSP6SetDataModule.sol [hunk 4]
```diff
@@ -228,6 +229,16 @@ abstract contract LSP6SetDataModule {
                 bytes12(inputDataKey) ==
                 _LSP6KEY_ADDRESSPERMISSIONS_PERMISSIONS_PREFIX
             ) {
+                // validate Data Value length
```
Comment (CJ42) [inline contracts/LSP6KeyManager/LSP6Modules/LSP6SetDataModule.sol:232]:
```suggestion
                // validate the number of bytes in `dataValue`
```

File: contracts/LSP6KeyManager/LSP6Modules/LSP6SetDataModule.sol [hunk 5]
```diff
@@ -286,6 +297,16 @@ abstract contract LSP6SetDataModule {
             inputDataKey == _LSP1_UNIVERSAL_RECEIVER_DELEGATE_KEY ||
             bytes12(inputDataKey) == _LSP1_UNIVERSAL_RECEIVER_DELEGATE_PREFIX
         ) {
+            // validate Data Value length
+            if (inputDataValue.length != 20 && inputDataValue.length != 0) {
+                revert InvalidDataValuesForDataKeys(
+                    inputDataKey,
+                    inputDataValue,
+                    0,
```
Comment (CJ42) [inline contracts/LSP6KeyManager/LSP6Modules/LSP6SetDataModule.sol:305]: I think we can get rid of this parameter.
And just replace the second one by the `invalidLength` of the `byte memory dataValue`

File: docs/contracts/LSP6KeyManager/LSP6KeyManager.md
Comment (CJ42) [inline docs/contracts/LSP6KeyManager/LSP6KeyManager.md:1]: Nice! 🙏

File: tests/LSP20CallVerification/LSP6/Admin/PermissionChangeAddExtensions.test.ts [hunk 6]
```diff
@@ -552,7 +552,7 @@ export const shouldBehaveLikePermissionChangeOrAddExtensions = (
             ],
             dataValues: [
               extensionB,
-              ethers.utils.hexZeroPad(ethers.utils.hexlify(8), 32),
+              ethers.utils.hexZeroPad(ethers.utils.hexlify(8), 16),
```
Comment (CJ42) [inline tests/LSP20CallVerification/LSP6/Admin/PermissionChangeAddExtensions.test.ts:555]: Nice catch! 🙏

**Review by CJ42 — CHANGES_REQUESTED** [review]

**Review by CJ42 — CHANGES_REQUESTED** [review]
Let's please review the newly added errors. This is adding 66k gas on deployment. It's way too much for just one check and two new custom errors.

Let's see if we can reduce this gas by some small refactoring or re-using the other errors 🙏

File: contracts/LSP6KeyManager/LSP6Modules/LSP6SetDataModule.sol (hunk 5 above)
Comment (b00ste) [reply contracts/LSP6KeyManager/LSP6Modules/LSP6SetDataModule.sol:305]: Wdym?

File: contracts/LSP6KeyManager/LSP6Errors.sol [hunk 7]
```diff
@@ -242,6 +230,17 @@ error RelayCallBeforeStartTime();
...
 error KeyManagerCannotBeSetAsExtensionForLSP20Functions();
+
+/**
+ * @notice Data value: `dataValue` length is different from the required length for the data key which is set.
+ *
+ * @dev Reverts when the data value length is not one of the required lengths for the specific data key.
+ *
+ * @param dataValue The data value that has an invalid length.
+ */
+error InvalidDataValuesForDataKeys(bytes dataValue);
```
Comment (YamenMerhi) [inline contracts/LSP6KeyManager/LSP6Errors.sol:246]: Maybe add the data key as parameter to know that the dataValue is invalid for what data key

**Review by YamenMerhi — COMMENTED** [review]
Was it discussed to have it strictly the length or the minimum length?

**Comment by CJ42** [comment]
> Was it discussed to have it strictly the length or the minimum length?

Good point. I remember @frozeman  saying it should be at least the minimum length. Any extra byte we ignore it.

**Review by skimaharvey — APPROVED** [review]
LGTM

**Comment by skimaharvey** [comment]
> > Was it discussed to have it strictly the length or the minimum length?
> 
> Good point. I remember @frozeman saying it should be at least the minimum length. Any extra byte we ignore it.
I believe for most strictly makes more sense as we are casting after but could be wrong

File: contracts/LSP6KeyManager/LSP6Modules/LSP6SetDataModule.sol [hunk 8]
```diff
@@ -228,6 +228,14 @@ abstract contract LSP6SetDataModule {
                 bytes12(inputDataKey) ==
                 _LSP6KEY_ADDRESSPERMISSIONS_PERMISSIONS_PREFIX
             ) {
+                // validate the number of bytes in `dataValue`
```
Comment (CJ42) [inline contracts/LSP6KeyManager/LSP6Modules/LSP6SetDataModule.sol:231]:
```suggestion
                // CHECK that `dataValue` contains at least 32 bytes, which is the required length for a permission BitArray
```

File: contracts/LSP6KeyManager/LSP6Modules/LSP6SetDataModule.sol [hunk 9]
```diff
@@ -286,6 +294,14 @@ abstract contract LSP6SetDataModule {
             inputDataKey == _LSP1_UNIVERSAL_RECEIVER_DELEGATE_KEY ||
             bytes12(inputDataKey) == _LSP1_UNIVERSAL_RECEIVER_DELEGATE_PREFIX
         ) {
+            // validate the number of bytes in `dataValue`
```
Comment (CJ42) [inline contracts/LSP6KeyManager/LSP6Modules/LSP6SetDataModule.sol:297]:
```suggestion
            // CHECK that `dataValue` contains at least 20 bytes, which corresponds to an address for a LSP1 Delegate contract
```

File: contracts/LSP6KeyManager/LSP6Modules/LSP6SetDataModule.sol [hunk 10]
```diff
@@ -305,6 +321,14 @@ abstract contract LSP6SetDataModule {
 
             // LSP17Extension:<bytes4>
         } else if (bytes12(inputDataKey) == _LSP17_EXTENSION_PREFIX) {
+            // validate the number of bytes in `dataValue`
```
Comment (CJ42) [inline contracts/LSP6KeyManager/LSP6Modules/LSP6SetDataModule.sol:324]:
```suggestion
            // CHECK that `dataValue` contains at least 20 bytes, which corresponds to an address for a LSP17 Extension
```

File: contracts/LSP6KeyManager/LSP6Modules/LSP6SetDataModule.sol [hunk 11]
```diff
@@ -355,6 +379,14 @@ abstract contract LSP6SetDataModule {
     ) internal view virtual returns (bytes32) {
         // AddressPermissions[] -> array length
         if (inputDataKey == _LSP6KEY_ADDRESSPERMISSIONS_ARRAY) {
+            // validate the number of bytes in `dataValue`
+            if (inputDataValue.length != 16 && inputDataValue.length != 0) {
```
Comment (CJ42) [inline contracts/LSP6KeyManager/LSP6Modules/LSP6SetDataModule.sol:383]: For this I would keep exactly like this, to be exactly 16 bytes long or 0 bytes, as this is how we check it also in LSP5Utils.

https://github.com/lukso-network/lsp-smart-contracts/blob/083d9970065b747ce0e1bfb5f4d98a3a86a164f4/contracts/LSP5ReceivedAssets/LSP5Utils.sol#L80-L87

File: contracts/LSP6KeyManager/LSP6Modules/LSP6SetDataModule.sol [hunk 12]
```diff
@@ -355,6 +379,14 @@ abstract contract LSP6SetDataModule {
     ) internal view virtual returns (bytes32) {
         // AddressPermissions[] -> array length
         if (inputDataKey == _LSP6KEY_ADDRESSPERMISSIONS_ARRAY) {
+            // validate the number of bytes in `dataValue`
```
Comment (CJ42) [inline contracts/LSP6KeyManager/LSP6Modules/LSP6SetDataModule.sol:382]:
```suggestion
            // CHECK that `dataValue` is exactly 16 bytes long or `0x` (= not set), as the array length of `AddressPermissions[]` MUST be a `uint128` value.
```

**Review by CJ42 — COMMENTED** [review]
Added some final suggestions for the Natspec

**Review by CJ42 — APPROVED** [review]

**Pull request merged**